### PR TITLE
Load cross module information during a meet

### PR DESCRIPTION
**Pull request by chambart**
When building an union approximation for the result of multiple branches, if some branch returns a value from an other compilation unit, it may fail to build an useful approximation. An example of a missed optimization opportunity is provided in the testsuite.

### CODE REVIEW

**Comment by mshinwell** [comment]
I can review this when I'm back at the start of next week.

**Comment by gasche** [comment]
I think that depending on the `backend` value rather than the specific `really_import_approx` function could make the caller-side code less repetitive (especially if you used the `E(nv).backend` accessor) and would also make sense on the function-definition side -- maybe later refinements could use other backend-dependent informations.

**Comment by chambart** [comment]
@gasche It's sadly not possible since Backend_intf depend on Simple_value_approx. This would require moving meet out of it. In particular, really_import_approx returns an Simple_value_approx.t.
But it is still possible to factorize a bit by having an `env => really_import_approx` function.

File: testsuite/tests/asmcomp/is_static_flambda.ml [hunk 1]
```diff
@@ -91,3 +91,26 @@ let () =
     try (failwith [@inlined always]) "some other string" with exn -> exn
   in
   assert(is_in_static_data exn)
+
+(* Verify that approximation intersection correctly loads exported
+   approximations.
+
+   Is_static_flambda_dep.cpl is a couple with 1 as first element. The
```
Comment (mshinwell) [inline testsuite/tests/asmcomp/is_static_flambda.ml]: "Pair" not "couple"

File: testsuite/tests/asmcomp/is_static_flambda.ml [hunk 2]
```diff
@@ -91,3 +91,26 @@ let () =
...
+   intersection of approximations should return a block of size 2 and
+   tag 0 containing 1 in the first field.
```
Comment (mshinwell) [inline testsuite/tests/asmcomp/is_static_flambda.ml]: I'm not sure this description is quite right, although maybe it's just the wording being confusing.  The intersection of approximations yields `[tag 0: [tag 0: Int 1, Unknown], Unknown]`, no?

File: middle_end/inline_and_simplify_aux.mli [hunk 3]
```diff
@@ -39,6 +39,10 @@ module Env : sig
       compiler backend being used for compilation. *)
   val backend : t -> (module Backend_intf.S)
 
+  (** Obtain the really_import_approx function from the backend module. *)
+  val really_import_approx : t ->
+    (Simple_value_approx.t -> Simple_value_approx.t)
+
```
Comment (mshinwell) [inline middle_end/inline_and_simplify_aux.mli:46]: Please reformat (cf. `create`)

File: middle_end/inline_and_simplify.ml [hunk 4]
```diff
@@ -1307,28 +1311,35 @@ and simplify env r (tree : Flambda.t) : Flambda.t * R.t =
...
     simplify_free_variable env arg ~f:(fun env arg _arg_approx ->
+      let really_import_approx = E.really_import_approx env in
+      let env = E.inside_branch env in
       let sw, r =
         List.fold_right (fun (str, lam) (sw, r) ->
+            let approx = R.approx r in
             let lam, r = simplify env r lam in
-            (str, lam)::sw, r)
+            (str, lam)::sw,
+            R.set_approx r (A.meet ~really_import_approx (R.approx r) approx))
```
Comment (mshinwell) [inline middle_end/inline_and_simplify.ml]: What about adding `R.meet_approx`?  It might simplify the code in these parts a bit.

**Comment by mshinwell** [comment]
I made a few comments.  I think it's ok apart from those.

File: middle_end/inline_and_simplify.ml [hunk 5]
```diff
@@ -1293,10 +1295,12 @@ and simplify env r (tree : Flambda.t) : Flambda.t * R.t =
         lam, R.map_benefit r B.remove_branch
       | _ ->
         let env = E.inside_branch env in
+        let really_import_approx = E.really_import_approx env in
         let f (i, v) (acc, r) =
           let approx = R.approx r in
           let lam, r = simplify env r v in
-          ((i, lam)::acc, R.set_approx r (A.meet (R.approx r) approx))
+          ((i, lam)::acc,
+           R.set_approx r (A.meet ~really_import_approx (R.approx r) approx))
```
Comment (mshinwell) [inline middle_end/inline_and_simplify.ml]: Needs another space of indentation (and below, please indent by 2 spaces when "Some l, R.set_approx ..." is split across two lines).

**Comment by chambart** [comment]
Thanks for the review.
By the way @mshinwell, do you have an .ocp-indent corresponding to this style. I assume that this is quite close to JS internal one.

**Comment by mshinwell** [comment]
I do not...

**Comment by mshinwell** [comment]
I think this is ok to push once those comments have been addressed.

**Comment by chambart** [comment]
I'll wait for CI to complete and push.

**Pull request merged**